### ISSUE TITLE
Scylla rejects deletions with empty-tuple restrictions

### ISSUE THREAD

**Issue by dekimir** [scylladb/scylla#9311]
This is an unintended consequence of 1fdaeca7d.

Scylla:
```cql
cqlsh:ks> CREATE TABLE test (k1 int, k2 int, v int, PRIMARY KEY (k1, k2));
cqlsh:ks> DELETE FROM test WHERE k1 IN ();
InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid partition key in a modification statement"
```

Cassandra:
```cql
cqlsh:ks> CREATE TABLE test (k1 int, k2 int, v int, PRIMARY KEY (k1, k2));
cqlsh:ks> DELETE FROM test WHERE k1 IN ();
```

**Comment by bhalevy** [comment]
@scylladb/scylla-maint we hit this issue in https://jenkins.scylladb.com/job/scylla-4.4/job/dtest-release/121/testReport/cql_additional_tests/TestCQL/empty_in_test/

Since branch-4.4 now includes 146f7b5421a2a8f5e7e7f531db78ca53c2cb4eec

Please backport

**Comment by psarna** [comment]
will do

**Comment by psarna** [comment]
update: very non-trivial backport, it will take me a while, since it's no longer possible to ask the author to try and do so

**Comment by bhalevy** [comment]
@psarna  Then maybe we should revert 146f7b5

**Comment by psarna** [comment]
we could, but are empty tuple restrictions ever seen outside of our test environment? Perhaps it's better to keep the patch that rejects updates with null key values at the price of not allowing empty tuple restrictions. Unless some 4.4 users actually rely on empty tuple restrictions, in which case it's better to revert

**Comment by nyh** [comment]
@psarna so you're suggesting that for 4.4, we need to disable the *test*, and leave the code as it is now?

**Comment by psarna** [comment]
@nyh yes, that's one of the ways to deal with the problem, but it depends on at least two important factors:
1. Are any of our users actually hit by this problem? If there are any, we should consider either a backport or a revert
2. How long until 4.4 reaches end-of-life? Doing a complicated backport for the sake of fixing a release for 2 weeks is probably not worth it

**Comment by avikivity** [comment]
It's not an important bug.

How did it start happening? Was a new test added?

**Comment by aleksbykov** [comment]
Now, rolling upgrade stay the same. Just for v4.4.8 upgrade starts failing once first node was upgrade and these query failed with error instead of return empty results, Also i think the upgrade tests from 4.4 -> 4.5  could fail before upgrade because same query executed before upgrade

**Comment by avikivity** [comment]
I reverted it from 4.4, we can revisit it but this gets us back to status quo ante.

**Comment by bhalevy** [comment]
@piotr how about 4.5?
It has 44c784cb79a698089c72f8db5907ad1906e0348b
but not 6afdc6004c2871f2a941276a9780d743a974d9f7

**Comment by bhalevy** [comment]
> how about 4.5?
> It has [44c784c](https://github.com/scylladb/scylla/commit/44c784cb79a698089c72f8db5907ad1906e0348b)
> but not [6afdc60](https://github.com/scylladb/scylla/commit/6afdc6004c2871f2a941276a9780d743a974d9f7)

@psarna ^^ (wrong user tagged previously, sorry)

**Comment by bhalevy** [comment]
And this regression since https://jenkins.scylladb.com/view/scylla-4.5/job/scylla-4.5/job/dtest-release/111/testReport/cql_additional_tests/TestCQL/empty_in_test/ 
at 44c784cb79a698089c72f8db5907ad1906e0348b

@avikivity so we should probably revert 44c784cb79a698089c72f8db5907ad1906e0348b from 4.5

**Comment by avikivity** [comment]
Reverted from 4.5.

**Comment by fgelcer** [comment]
> Reverted from 4.5.

now all rolling upgrade jobs from 4.4 and 4.5 to 4.5.3 are failing with this error:
```
DELETE FROM empty_in_test1 WHERE k1 IN () < t:2021-12-22 04:26:51,359 f:fill_db_data.py l:3220 c:sdcm.fill_db_data    p:ERROR > DELETE FROM empty_in_test1 WHERE k1 IN ()
Traceback (most recent call last):
  File "/jenkins/slave/workspace/scylla-4.5/rolling-upgrade/rolling-upgrade-centos7-test@2/scylla-cluster-tests/sdcm/fill_db_data.py", line 3217, in _run_db_queries
    res = session.execute(item['queries'][i])
  File "/jenkins/slave/workspace/scylla-4.5/rolling-upgrade/rolling-upgrade-centos7-test@2/scylla-cluster-tests/sdcm/utils/common.py", line 1263, in execute_verbose
    return execute_orig(*args, **kwargs)
  File "cassandra/cluster.py", line 2611, in cassandra.cluster.Session.execute
  File "cassandra/cluster.py", line 4829, in cassandra.cluster.ResponseFuture.result
cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Invalid partition key in a modification statement"
```

it means that, once these changes were reverted for 4.5 (and 4.4) the tests in SCT also had to be changed to not fail with this empty tuple...

is it still supported for 4.6?

**Comment by avikivity** [comment]
The bug should be fixed in 4.6 (and not in 4.4 and 4.5, after the reverts).
